File: cpp_generated/errors.test.cpp
#include "/home/runner/work/elizaos-cpp/elizaos-cpp/classified/packages/plugin-stagehand/src/__tests__/errors.test.h"

void Main(void)
{
    describe(std::string("Error Classes"), [=]() mutable
    {
        describe(std::string("StagehandError"), [=]() mutable
        {
            it(std::string("should create error with all properties"), [=]() mutable
            {
                auto error = std::make_shared<StagehandError>(std::string("Test error"), std::string("TEST_ERROR"), std::string("User friendly message"), true, object{
                    object::pair{std::string("extra"), std::string("data")}
                });
                expect(error->message)->toBe(std::string("Test error"));
                expect(error->code)->toBe(std::string("TEST_ERROR"));
                expect(error->userMessage)->toBe(std::string("User friendly message"));
                expect(error->recoverable)->toBe(true);
                expect(error->details)->toEqual(object{
                    object::pair{std::string("extra"), std::string("data")}
                });
                expect(error->name)->toBe(std::string("StagehandError"));
            }
            );
            it(std::string("should default recoverable to true"), [=]() mutable
            {
                auto error = std::make_shared<StagehandError>(std::string("Test error"), std::string("TEST_ERROR"), std::string("User friendly message"));
                expect(error->recoverable)->toBe(true);
            }
            );
        }
        );
        describe(std::string("BrowserNavigationError"), [=]() mutable
        {
            it(std::string("should create navigation error with URL"), [=]() mutable
            {
                auto error = std::make_shared<BrowserNavigationError>(std::string("https://example.com"));
                expect(error->code)->toBe(std::string("NAVIGATION_ERROR"));
                expect(error->message)->toContain(std::string("Failed to navigate to https://example.com"));
                expect(error->userMessage)->toContain(std::string("couldn't navigate to the requested page"));
                expect(error->recoverable)->toBe(true);
                expect(error->details["url"])->toBe(std::string("https://example.com"));
            }
            );
            it(std::string("should include original error message"), [=]() mutable
            {
                auto originalError = std::make_shared<Error>(std::string("Network timeout"));
                auto error = std::make_shared<BrowserNavigationError>(std::string("https://example.com"), originalError);
                expect(error->message)->toContain(std::string("Network timeout"));
                expect(error->details["originalError"])->toBe(std::string("Network timeout"));
            }
            );
        }
        );
        describe(std::string("BrowserSessionError"), [=]() mutable
        {
            it(std::string("should create session error"), [=]() mutable
            {
                auto error = std::make_shared<BrowserSessionError>(std::string("Session expired"), object{
                    object::pair{std::string("sessionId"), std::string("123")}
                });
                expect(error->code)->toBe(std::string("BROWSER_SESSION_ERROR"));
                expect(error->message)->toBe(std::string("Session expired"));
                expect(error->userMessage)->toContain(std::string("There was an error with the browser session"));
                expect(error->details["sessionId"])->toBe(std::string("123"));
            }
            );
        }
        );
        describe(std::string("BrowserActionError"), [=]() mutable
        {
            it(std::string("should create action error"), [=]() mutable
            {
                auto error = std::make_shared<BrowserActionError>(std::string("click"), std::string("submit button"));
                expect(error->code)->toBe(std::string("ACTION_ERROR"));
                expect(error->message)->toContain(std::string("Failed to click on submit button"));
                expect(error->userMessage)->toContain(std::string("I couldn't click on the requested element"));
                expect(error->details["action"])->toBe(std::string("click"));
                expect(error->details["target"])->toBe(std::string("submit button"));
            }
            );
            it(std::string("should include original error"), [=]() mutable
            {
                auto originalError = std::make_shared<Error>(std::string("Element not found"));
                auto error = std::make_shared<BrowserActionError>(std::string("type"), std::string("input field"), originalError);
                expect(error->message)->toContain(std::string("Element not found"));
                expect(error->details["originalError"])->toBe(std::string("Element not found"));
            }
            );
        }
        );
        describe(std::string("BrowserSecurityError"), [=]() mutable
        {
            it(std::string("should create security error"), [=]() mutable
            {
                auto error = std::make_shared<BrowserSecurityError>(std::string("Blocked domain"), object{
                    object::pair{std::string("domain"), std::string("malware.com")}
                });
                expect(error->code)->toBe(std::string("SECURITY_ERROR"));
                expect(error->message)->toBe(std::string("Blocked domain"));
                expect(error->userMessage)->toContain(std::string("blocked for security reasons"));
                expect(error->recoverable)->toBe(false);
                expect(error->details["domain"])->toBe(std::string("malware.com"));
            }
            );
        }
        );
        describe(std::string("BrowserServiceNotAvailableError"), [=]() mutable
        {
            it(std::string("should create service not available error"), [=]() mutable
            {
                auto error = std::make_shared<BrowserServiceNotAvailableError>();
                expect(error->code)->toBe(std::string("SERVICE_NOT_AVAILABLE"));
                expect(error->message)->toBe(std::string("Browser service is not available"));
                expect(error->userMessage)->toContain(std::string("The browser automation service is not available"));
                expect(error->recoverable)->toBe(false);
            }
            );
        }
        );
    }
    );
    describe(std::string("handleBrowserError"), [=]() mutable
    {
        it(std::string("should handle StagehandError"), [=]() mutable
        {
            auto callback = mock();
            auto error = std::make_shared<BrowserNavigationError>(std::string("https://example.com"));
            handleBrowserError(error, callback);
            expect(callback)->toHaveBeenCalledWith(object{
                object::pair{std::string("text"), error->userMessage}, 
                object::pair{std::string("error"), true}
            });
        }
        );
        it(std::string("should convert timeout errors"), [=]() mutable
        {
            auto callback = mock();
            auto error = std::make_shared<Error>(std::string("Operation timeout after 5000ms"));
            handleBrowserError(error, callback, std::string("load page"));
            expect(callback)->toHaveBeenCalledWith(object{
                object::pair{std::string("text"), std::string("I encountered an error while trying to load page. Please try again.")}, 
                object::pair{std::string("error"), true}
            });
        }
        );
        it(std::string("should convert navigation errors"), [=]() mutable
        {
            auto callback = mock();
            auto error = std::make_shared<Error>(std::string("Failed to navigate to page"));
            handleBrowserError(error, callback);
            expect(callback)->toHaveBeenCalledWith(object{
                object::pair{std::string("text"), std::string("I encountered an unexpected error. Please try again.")}, 
                object::pair{std::string("error"), true}
            });
        }
        );
        it(std::string("should handle generic errors"), [=]() mutable
        {
            auto callback = mock();
            auto error = std::make_shared<Error>(std::string("Something went wrong"));
            handleBrowserError(error, callback, std::string("perform action"));
            expect(callback)->toHaveBeenCalledWith(object{
                object::pair{std::string("text"), std::string("I encountered an error while trying to perform action. Please try again.")}, 
                object::pair{std::string("error"), true}
            });
        }
        );
        it(std::string("should handle non-Error objects"), [=]() mutable
        {
            auto callback = mock();
            auto error = std::make_shared<Error>(std::string("String error"));
            handleBrowserError(error, callback, std::string("do something"));
            expect(callback)->toHaveBeenCalledWith(object{
                object::pair{std::string("text"), std::string("I encountered an error while trying to do something. Please try again.")}, 
                object::pair{std::string("error"), true}
            });
        }
        );
        it(std::string("should use default action when not provided"), [=]() mutable
        {
            auto callback = mock();
            auto error = std::make_shared<Error>(std::string("Test error"));
            handleBrowserError(error, callback);
            expect(callback)->toHaveBeenCalledWith(object{
                object::pair{std::string("text"), std::string("I encountered an unexpected error. Please try again.")}, 
                object::pair{std::string("error"), true}
            });
        }
        );
    }
    );
}

MAIN

File: classified/packages/plugin-stagehand/src/__tests__/errors.test.ts
import { describe, expect, it, mock } from 'bun:test';
import {
  StagehandError,
  BrowserNavigationError,
  BrowserSessionError,
  BrowserActionError,
  BrowserSecurityError,
  BrowserServiceNotAvailableError,
  handleBrowserError,
} from '../errors';

describe('Error Classes', () => {
  describe('StagehandError', () => {
    it('should create error with all properties', () => {
      const error = new StagehandError('Test error', 'TEST_ERROR', 'User friendly message', true, {
        extra: 'data',
      });

      expect(error.message).toBe('Test error');
      expect(error.code).toBe('TEST_ERROR');
      expect(error.userMessage).toBe('User friendly message');
      expect(error.recoverable).toBe(true);
      expect(error.details).toEqual({ extra: 'data' });
      expect(error.name).toBe('StagehandError');
    });

    it('should default recoverable to true', () => {
      const error = new StagehandError('Test error', 'TEST_ERROR', 'User friendly message');

      expect(error.recoverable).toBe(true);
    });
  });

  describe('BrowserNavigationError', () => {
    it('should create navigation error with URL', () => {
      const error = new BrowserNavigationError('https://example.com');

      expect(error.code).toBe('NAVIGATION_ERROR');
      expect(error.message).toContain('Failed to navigate to https://example.com');
      expect(error.userMessage).toContain("couldn't navigate to the requested page");
      expect(error.recoverable).toBe(true);
      expect(error.details?.url).toBe('https://example.com');
    });

    it('should include original error message', () => {
      const originalError = new Error('Network timeout');
      const error = new BrowserNavigationError('https://example.com', originalError);

      expect(error.message).toContain('Network timeout');
      expect(error.details?.originalError).toBe('Network timeout');
    });
  });

  describe('BrowserSessionError', () => {
    it('should create session error', () => {
      const error = new BrowserSessionError('Session expired', { sessionId: '123' });

      expect(error.code).toBe('BROWSER_SESSION_ERROR');
      expect(error.message).toBe('Session expired');
      expect(error.userMessage).toContain('There was an error with the browser session');
      expect(error.details?.sessionId).toBe('123');
    });
  });

  describe('BrowserActionError', () => {
    it('should create action error', () => {
      const error = new BrowserActionError('click', 'submit button');

      expect(error.code).toBe('ACTION_ERROR');
      expect(error.message).toContain('Failed to click on submit button');
      expect(error.userMessage).toContain("I couldn't click on the requested element");
      expect(error.details?.action).toBe('click');
      expect(error.details?.target).toBe('submit button');
    });

    it('should include original error', () => {
      const originalError = new Error('Element not found');
      const error = new BrowserActionError('type', 'input field', originalError);

      expect(error.message).toContain('Element not found');
      expect(error.details?.originalError).toBe('Element not found');
    });
  });

  // TODO: Implement BrowserTimeoutError
  // describe('BrowserTimeoutError', () => {
  //   it('should create timeout error', () => {
  //     const error = new BrowserTimeoutError('navigation', 30000);

  //     expect(error.code).toBe('BROWSER_TIMEOUT_ERROR');
  //     expect(error.message).toContain('timed out after 30000ms');
  //     expect(error.userMessage).toContain('took too long and timed out');
  //     expect(error.details?.action).toBe('navigation');
  //     expect(error.details?.timeoutMs).toBe(30000);
  //   });
  // });

  // TODO: Implement BrowserExtractionError
  // describe('BrowserExtractionError', () => {
  //   it('should create extraction error', () => {
  //     const error = new BrowserExtractionError('Extract user name');

  //     expect(error.code).toBe('BROWSER_EXTRACTION_ERROR');
  //     expect(error.message).toContain('Failed to extract data');
  //     expect(error.userMessage).toContain("couldn't extract the requested information");
  //     expect(error.details?.instruction).toBe('Extract user name');
  //   });

  //   it('should include original error', () => {
  //     const originalError = new Error('Invalid selector');
  //     const error = new BrowserExtractionError('Extract price', originalError);

  //     expect(error.message).toContain('Invalid selector');
  //     expect(error.details?.originalError).toBe('Invalid selector');
  //   });
  // });

  describe('BrowserSecurityError', () => {
    it('should create security error', () => {
      const error = new BrowserSecurityError('Blocked domain', { domain: 'malware.com' });

      expect(error.code).toBe('SECURITY_ERROR');
      expect(error.message).toBe('Blocked domain');
      expect(error.userMessage).toContain('blocked for security reasons');
      expect(error.recoverable).toBe(false);
      expect(error.details?.domain).toBe('malware.com');
    });
  });

  describe('BrowserServiceNotAvailableError', () => {
    it('should create service not available error', () => {
      const error = new BrowserServiceNotAvailableError();

      expect(error.code).toBe('SERVICE_NOT_AVAILABLE');
      expect(error.message).toBe('Browser service is not available');
      expect(error.userMessage).toContain('The browser automation service is not available');
      expect(error.recoverable).toBe(false);
    });
  });
});

describe('handleBrowserError', () => {
  it('should handle StagehandError', () => {
    const callback = mock();
    const error = new BrowserNavigationError('https://example.com');

    handleBrowserError(error, callback);

    expect(callback).toHaveBeenCalledWith({
      text: error.userMessage,
      error: true,
    });
  });

  it('should convert timeout errors', () => {
    const callback = mock();
    const error = new Error('Operation timeout after 5000ms');

    handleBrowserError(error, callback, 'load page');

    expect(callback).toHaveBeenCalledWith({
      text: 'I encountered an error while trying to load page. Please try again.',
      error: true,
    });
  });

  it('should convert navigation errors', () => {
    const callback = mock();
    const error = new Error('Failed to navigate to page');

    handleBrowserError(error, callback);

    expect(callback).toHaveBeenCalledWith({
      text: 'I encountered an unexpected error. Please try again.',
      error: true,
    });
  });

  it('should handle generic errors', () => {
    const callback = mock();
    const error = new Error('Something went wrong');

    handleBrowserError(error, callback, 'perform action');

    expect(callback).toHaveBeenCalledWith({
      text: 'I encountered an error while trying to perform action. Please try again.',
      error: true,
    });
  });

  it('should handle non-Error objects', () => {
    const callback = mock();
    const error = new Error('String error');

    handleBrowserError(error, callback, 'do something');

    expect(callback).toHaveBeenCalledWith({
      text: 'I encountered an error while trying to do something. Please try again.',
      error: true,
    });
  });

  it('should use default action when not provided', () => {
    const callback = mock();
    const error = new Error('Test error');

    handleBrowserError(error, callback);

    expect(callback).toHaveBeenCalledWith({
      text: 'I encountered an unexpected error. Please try again.',
      error: true,
    });
  });
});
